### ISSUE TITLE
Extract CSS separately for every file of the array of rollup entry points

### ISSUE THREAD

**Issue by toli444** [egoist/rollup-plugin-postcss#160]
Extracting CSS into a single file breaks treeshaking. I only want to ship critical styles. So as rollup turns every entry point to separate output chunks it would be great if there were a possibility to have the same for CSS instead of just generating one file which contains all the styles.

// config
```
rollupConfig = {
    input: ['./src/components/Button/index.js', './src/components/Input/index.js'],
    output: [
        {
            dir: lib,
            format: 'cjs',
            sourcemap: true,
            exports: 'named'
        },
    ],
    plugins: [
        postcss({
                modules: true,
                extract: true,
                sourceMap: true,
                minimize: true
            })
    ]
}
```
// result
```
.
└───components
│   └─── Input
│       │   index.js
│   └─── Button
│       │   index.js
└───folder2
│   index.js
│   chunk.js
│   styles.css
```

//expected result
```
//result
```
.
└───components
│   └─── Input
│       │   index.js
│       │   styles.css
│   └─── Button
│       │   index.js
│       │   styles.css
└───folder2
│   index.js
│   chunk.js
│   styles.css
```

**Comment by tianyingchun** [comment]
has any progress for  this feature ?

**Comment by krvajal** [comment]
This is what I am doing
```

const entryPoints = {
  entry1: 'src/entry1.js',
  entry2: 'src/entry2.js',
}

// this needs to be a different instance every time
function getPlugins(){
	return [postcss()]
}

const configs = Object.keys(entryPoints).reduce((configs, entry) => {
  const input = entryPoints[entry]

  const config = {
    input,
    output: {
      file: `build/${entry}.js`,
      format: 'esm',
      name: entry,
      globals,
    },
    plugins: getPlugins(),
    external,
  }
  configs.push(config)
  return configs
}, [])

module.exports = configs;

```

**Comment by mlnor27** [comment]
@krvajal This is a bad workaround as it repeats the bundle process for each entry + produces code duplication...

Is there anything planned for this matter anytime soon ? @egoist

**Comment by whatisaphone** [comment]
This is particularly important if you're bundling multiple apps that each contain their own styles for `html`, `body`, etc.

**Comment by frattaro** [comment]
Here's a use case:

You have a large react-rollup component library that is consumed by a `create-react-app` application. They only need one component, and also need to run postcss on that component's css to ensure it's properly scoped. The component library already ships each component as a separate entry point, but can't get the per-component css extracted, and are unable to support postcss in the consuming application.

**Comment by SasanFarrokh** [comment]
Could we do not extract css for our component library, but extract them in our final build website?, i mean somehow mark the css injection codes for later extraction?

**Comment by wedneyyuri** [comment]
Same problem here, I would like to break in small chunks to expose a components package which will be imported later using NextJS.

**Comment by hnrchrdl** [comment]
I also have this requirement. I would like my app to be able to import the library css separately for every component.

**Comment by kamaladenalhomsi** [comment]
well so far this is my workaround:

prerequisites:
```
yarn add -D glob or npm install --dev glob
```

```js
// rollup.config.js

const glob = require('glob')
const path = require('path')
const postcss = require('rollup-plugin-postcss')

const bundleCss = () => {
  var config = []
  var files = glob.sync(path.resolve(__dirname, '**/*.css'))
  files.forEach(file => {
    var filename = file.substr(file.lastIndexOf('/') + 1, file.length).toLowerCase()
    config.push(
      postcss({
        include: file,
        extract: path.resolve(`dist/${filename}`),
        minimize: true
      })
    )
  })
  return config
}

modules.exports = {
  plugins: [
    ...bundleCss()
  ]
}

```

basically I am looping over all css files in my project, and returning an array of postcss plugin for each file

**Comment by 3lang3** [comment]
has any progress for this feature ?

**Comment by laizp** [comment]
I have this requirement to.

**Comment by Brandonitas** [comment]
has any progress for this feature?

**Comment by iyinchao** [comment]
+1, this would be very handy

**Comment by DanielAmenou** [comment]
I created a plugin - [rollup-plugin-lib-style](https://www.npmjs.com/package/rollup-plugin-lib-style) that generates CSS separately for every file and imports these generated files, so the consumer of the package will not need to import each CSS file separately.
You can give it a try

**Comment by gde-pass** [comment]
Anything new about that ?

**Comment by uditalias** [comment]
has any progress for this feature?

**Comment by tianyingchun** [comment]
i pareparing create simular plugin to generates css modular support less, sass, postcss for rollup multi entries bundle at one time.

**Comment by tianyingchun** [comment]
@uditalias 
try `@flatjs/forge-plugin-styling` https://www.npmjs.com/package/@flatjs/forge-plugin-styling

<img width="456" alt="image" src="https://user-images.githubusercontent.com/2833251/216292619-1caad2aa-0fc4-40a4-ae49-0d683694c53c.png">
<img width="515" alt="image" src="https://user-images.githubusercontent.com/2833251/216292678-df0ed1c2-c7a7-46c3-a3de-2b5bb48a8bce.png">
<img width="536" alt="image" src="https://user-images.githubusercontent.com/2833251/216292711-fe69259b-e0b8-4677-9794-075ed128e51e.png">
<img width="547" alt="image" src="https://user-images.githubusercontent.com/2833251/216292743-1bab2758-1033-42fb-b713-6724a1b6b6ab.png">
<img width="477" alt="image" src="https://user-images.githubusercontent.com/2833251/216292770-40d557f2-1ccb-4a6b-bc89-3b79b49415e7.png">
<img width="663" alt="image" src="https://user-images.githubusercontent.com/2833251/216292884-4675b8de-6149-4eec-a89d-0cc9f8051d26.png">
<img width="677" alt="image" src="https://user-images.githubusercontent.com/2833251/216292945-c9edb84c-7bfa-4861-bea4-75c3de7c915d.png">
<img width="708" alt="image" src="https://user-images.githubusercontent.com/2833251/216292971-2d73ddbd-f27c-4d29-83a6-1d3e5901a6ac.png">

**Comment by tianyingchun** [comment]
BTW: rollup entries: 
```
 'src/module1/module1.ts',
          'src/module2/module2.ts',
          'src/module0/index.ts',
          'src/module-no-style/index.ts',
```

**Comment by cherylcarpenter** [comment]
Hoping we can keep this issue open.
My use case is that I am using rollup for a React Component Library. I want each js module to have it's own css sidecar.

**Comment by joemaller** [comment]
I'm looking to migrate a project out of webpack which should output one css file for each entry point. Output files are used in separate contexts, and styles should apply to each individually.

Here's a simplified example of what should happen: 

```
// main.js
import "main.scss";

// main.scss
body { color: red }

// admin.js
import "admin.scss"

// admin.scss
body { color: blue }
```

Output should look the same as above, except the scss files would be compiled and output as css. 

Instead, I get two js files and a single css file, with a filename based on the first entrypoint in rollup.config:
```
// main.css
body { color: blue } 
body { color: red }
```
(concatenation order appears to be bottom-up, but I'm not certain about that)

**Comment by SunHuawei** [comment]
> Hoping we can keep this issue open. My use case is that I am using rollup for a React Component Library. I want each js module to have it's own css sidecar.

Cannot agree more! My case is when using `preserveModules: true`, I expect it can keep the CSS references for each module(file) so that only the necessary CSS can be included in the final bundle when the end user builds with webpack.

**Comment by cristian-eriomenco** [comment]
Agree same problem. To add more, when using css modules with `extract: true`, ALL of your `<component>.module.css` gets aggregated in a dist/index.css. Now imagine importing only a Button component, the end result is unnecessary css being loaded, which doesn't quite align with tree-shaking/code-splitting concepts very well.

At the same time if you do `extract: false` then your resulting css is bundled as `<component>.module.css.js` alongside your `<component>.js` but this time you get degraded user experience caused by flickering, especially if this component is not/cannot be SSRed.

In summary at the moment its a choice between a large and heavy index.css or a split inlined css as js with degraded user experiece due to flickering / page shifts etc

**Comment by tmeindle** [comment]
I recently encountered similar issues.

I use preserveModules=true output option and  was trying to get assets to be included in the output folder in the correct locations relative to the importer in the output.  I was unsuccessful with every plugin I tried as they all want to bundle the with any of the styling plugins available or mentioned in this thread. 

In my use case I don't wish to process the stylesheets with postcss or sass.  I will defer transpiling and bundling of assets to the web app bundler (rsbuild in my case).  I just want to copy the imported assets next to the files that import them.

So to that end, I created a plugin that copies imported assets relative to the importer in the output.  It also scans css and scss for imports and includes assets from those in the output relative to the sheets that imports it.

It is available on npmjs in case anyone is interested:
https://www.npmjs.com/package/rollup-plugin-library-assets

**Comment by DarkGenius** [comment]
Has any progress for this feature?

**Comment by WindRunnerMax** [comment]
I have looked into similar issues https://github.com/egoist/rollup-plugin-postcss/issues/204, https://github.com/jleeson/rollup-plugin-import-css/issues/37 and https://github.com/egoist/rollup-plugin-postcss/issues/160 . Therefore, I attempted to resolve this problem by extending the `rollup-plugin-postcss` plugin. 

Parsing CSS still relies on `rollup-plugin-postcss`, and then I re-emit the content of each individually parsed file back into Rollup. This works well in version `4.0.2`.

Implementing this feature requires three steps:

Step 1: Copy the plugin's source code into your project. The code file is here: 

https://github.com/WindRunnerMax/BlockKit/blob/32d90028e90f8565c680f23ad0c93f60c41ce9ef/packages/plugin/script/postcss-preserve.js

Step 2: Replace the `postcss` plugin with the `postcssPreserve` plugin in your Rollup configuration.

https://github.com/WindRunnerMax/BlockKit/blob/32d90028e90f8565c680f23ad0c93f60c41ce9ef/packages/plugin/rollup.config.js#L60

```js
{
   // ...
   plugins: [
      // postcss( /** ... */ )
      postcssPreserve( /** ... */ )
   ]
}
```

Step 3: Explicitly pass the extract and preserveRoot parameters (typically pointing to your src directory).

https://github.com/WindRunnerMax/BlockKit/blob/32d90028e90f8565c680f23ad0c93f60c41ce9ef/packages/plugin/rollup.config.js#L60-L65

```js
postcssPreserve({
  preserveRoot: "./src",
  extract: "_.css",
  minimize: true,
  extensions: [".css", ".scss"],
})
```

Finally, you can refer to this npm package for the expected bundling result.

https://www.npmjs.com/package/@block-kit/plugin?activeTab=code
